Use testcontainers networking capabilities when using @QuarkusIntegrationTest + DevServices

geoand opened the pull request:
This is done in order to get us off of host networking (which only works on Linux).
The way this works is that when `@QuarkusIntegrationTest` is meant to launch a container
and DevServices are needed, a build item is pushed into the build that forces
the DevServices to launch in TestContainers' `SHARED` Network.
The id of network is then used when launching the application as a container.

This change only makes a difference for the container launch case of `@QuarkusIntegrationTest`.

Fixes: #18914

geoand committed:
Use testcontainers networking capabilities when using @QuarkusIntegrationTest + DevServices

This is done in order to get us off of host networking (which only works on Linux).
The way this works is that when @QuarkusIntegrationTest is meant to launch a container
and DevServices are needed, a build item is pushed into the build that forces
the DevServices to launch in TestContainers' SHARED Network.
The id of network is then used when launching the application as a container.

Fixes: #18194

geoand commented:
@stuartwdouglas mind taking a look to validate the approach before I move to update the rest of the DevServices?

stuartwdouglas commented:
I don't really like the complexity it adds to all the DevServices, although I also can't really think of any better options. 

Is there no way to have this logic in a common place somehow? It looks like it might be specific to each container, which means we really should test every DevServices in both shared and non-shared mode.

geoand commented:
> I don't really like the complexity it adds to all the DevServices, although I also can't really think of any better options.
> 

I agree and it's why I put this off for so long...

> Is there no way to have this logic in a common place somehow? It looks like it might be specific to each container, which means we really should test every DevServices in both shared and non-shared mode.

I don't think so as we need to manipulate the returned URL of each container. 
We could have the hostname generation be common, but that wouldn't really reduce the complexity much.

As for testing, we can convert some of the integration tests into container builds and use them.

geoand commented:
The PR should be ready to go now

sberyozkin commented:
Hi @geoand Can you please update `KeycloakDevServicesProcessor` as well ?

sberyozkin commented:
Sorry, you already did :-)

geoand commented:
> I updated all the DevServices and added a bunch of tests

I should also mention that these tests exercise both the shared network and default code paths because dev-services comes into play for them both for surefire and failsafe

geoand commented:
@stuartwdouglas can I get a review of this? Since it touches all the dev services, there is a good chance it could get conflicts soon

bdevreugd-vialis commented:
I do have one comment, because all the communication is now network internal, the issue of: https://github.com/quarkusio/quarkus/issues/18854 kind of returns.

I would still like to be able to access the dev services from the test code itself, maybe also expose a property with postfix `.from-host` that will contain the original URL (with the mapped port)?

geoand commented:
> I do have one comment, because all the communication is now network internal, the issue of: #18854 kind of returns.
> 
> I would still like to be able to access the dev services from the test code itself, maybe also expose a property with postfix `.from-host` that will contain the original URL (with the mapped port)?

Good point. Let's leave that for a next PR, so please open an issue once this goes in

geoand commented:
Ideally we would backport this one, but given the scope of the change I think it's safer not to, so I am removing the label

stuartwdouglas commented on extensions/devservices/db2/src/main/java/io/quarkus/devservices/db2/deployment/DB2DevServicesProcessor.java:
```diff
@@ -57,4 +53,53 @@ public void close() throws IOException {
+        protected void configure() {
+            super.configure();
+
+            if (useSharedNetwork) {
```
Is there any chance this code can be moved to a shared utility method so it is not copied everywhere?

geoand replied:
I'll look into it

geoand replied:
I moved the shared network stuff into a utility method. The rest isn't that straighforward as some of the methods are protected
